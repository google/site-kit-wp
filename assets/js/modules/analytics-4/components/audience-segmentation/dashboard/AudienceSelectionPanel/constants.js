/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
