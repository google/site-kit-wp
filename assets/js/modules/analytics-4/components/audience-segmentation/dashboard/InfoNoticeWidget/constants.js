/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
