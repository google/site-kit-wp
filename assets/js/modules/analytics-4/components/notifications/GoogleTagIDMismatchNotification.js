/**
 * GoogleTagIDMismatchNotification component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';
import { useCallback } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';
import { MODULE_SLUG_ANALYTICS_4 } from '../../constants';
import { isValidMeasurementID } from '../../utils/validation';
import { CORE_MODULES } from '../../../../googlesitekit/modules/datastore/constants';
import { getBestTagID } from '../../utils/google-tag';
import BannerNotification, {
	TYPES,
} from '@/js/googlesitekit/notifications/components/layout/BannerNotification';
import { ProgressBar } from 'googlesitekit-components';

export default function GoogleTagIDMismatchNotification( {
	id,
	Notification,
} ) {
	const currentAnalyticsConfig = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getSettings()
	);

	const googleTagContainerDestinations = useSelect(
		( select ) =>
			currentAnalyticsConfig?.googleTagAccountID &&
			currentAnalyticsConfig?.googleTagContainerID &&
			select( MODULES_ANALYTICS_4 ).getGoogleTagContainerDestinations(
				currentAnalyticsConfig.googleTagAccountID,
				currentAnalyticsConfig.googleTagContainerID
			)
	);

	const currentAnalyticsProperty = useSelect(
		( select ) =>
			currentAnalyticsConfig?.propertyID &&
			select( MODULES_ANALYTICS_4 ).getProperty(
				currentAnalyticsConfig.propertyID
			)
	);

	const newAnalyticsConfig = useSelect( ( select ) => {
		if ( ! Array.isArray( googleTagContainerDestinations ) ) {
			return null;
		}

		const destinationIDs = googleTagContainerDestinations.map(
			// eslint-disable-next-line sitekit/acronym-case
			( destination ) => destination.destinationId
		);

		const validGA4MeasurementIDs =
			destinationIDs.filter( isValidMeasurementID );

		if ( validGA4MeasurementIDs.length === 0 ) {
			return null;
		}

		return select( MODULES_ANALYTICS_4 ).getAnalyticsConfigByMeasurementIDs(
			validGA4MeasurementIDs
		);
	} );

	const newAnalyticsProperty = useSelect(
		( select ) =>
			newAnalyticsConfig &&
			newAnalyticsConfig.propertyID &&
			select( MODULES_ANALYTICS_4 ).getProperty(
				newAnalyticsConfig?.propertyID
			)
	);

	const newGoogleTagID = useSelect( ( select ) => {
		const tagIDs =
			currentAnalyticsConfig?.measurementID &&
			select( MODULES_ANALYTICS_4 ).getGoogleTagContainer(
				currentAnalyticsConfig?.measurementID
			)?.tagIds; // eslint-disable-line sitekit/acronym-case

		if ( Array.isArray( tagIDs ) ) {
			return getBestTagID( tagIDs, currentAnalyticsConfig.measurementID );
		}
	} );

	const isDoingSubmitChanges = useSelect( ( select ) =>
		select( CORE_MODULES ).isDoingSubmitChanges( MODULE_SLUG_ANALYTICS_4 )
	);

	const {
		setPropertyID,
		setWebDataStreamID,
		setMeasurementID,
		updateSettingsForMeasurementID,
		submitChanges,
		setHasMismatchedGoogleTagID,
	} = useDispatch( MODULES_ANALYTICS_4 );

	const updateToNewAnalyticsConfig = useCallback( async () => {
		setPropertyID( newAnalyticsConfig?.propertyID );
		setWebDataStreamID( newAnalyticsConfig?.webDataStreamID );
		setMeasurementID( newAnalyticsConfig?.measurementID );
		await submitChanges();
		setHasMismatchedGoogleTagID( false );
	}, [
		setPropertyID,
		setWebDataStreamID,
		setMeasurementID,
		newAnalyticsConfig?.propertyID,
		newAnalyticsConfig?.webDataStreamID,
		newAnalyticsConfig?.measurementID,
		submitChanges,
		setHasMismatchedGoogleTagID,
	] );

	const updateGoogleTagConfig = useCallback( async () => {
		await updateSettingsForMeasurementID(
			currentAnalyticsConfig?.measurementID
		);
		await submitChanges();
		setHasMismatchedGoogleTagID( false );
	}, [
		updateSettingsForMeasurementID,
		currentAnalyticsConfig?.measurementID,
		submitChanges,
		setHasMismatchedGoogleTagID,
	] );

	if (
		googleTagContainerDestinations === undefined ||
		newAnalyticsProperty === undefined ||
		newGoogleTagID === undefined ||
		currentAnalyticsProperty === undefined
	) {
		return (
			<ProgressBar
				className="googlesitekit-banner__progress-bar"
				height={ 4 } // The progress bar height is set to 7px via CSS. This value removes margins within the ProgressBar component.
				indeterminate
			/>
		);
	}

	// If the current and new properties are the same, don't show the notification.
	// This can happen momentarily when the banner is actioned and the new property
	// has been saved, but the notification hasn't been dismissed / removed from the
	// queue yet. There is also a rare issue reported where hasMismatchedGoogleTagID
	// is set to true, but the new property is the same as the current one. This will
	// result in a perpetual loading banner on the screen if we returned the loading
	// bar. So it is safer to return null, as eventually, the syncGoogleTag action
	// will be called within an hour and reset this correctly.
	if (
		currentAnalyticsProperty?._id &&
		newAnalyticsProperty?._id &&
		currentAnalyticsProperty._id === newAnalyticsProperty._id
	) {
		// return null;
	}

	if ( newAnalyticsProperty ) {
		return (
			<Notification className="googlesitekit-publisher-win">
				<BannerNotification
					notificationID={ id }
					type={ TYPES.WARNING }
					title={ __(
						"Update Site Kit's Analytics configuration to continue seeing Analytics data on the dashboard",
						'google-site-kit'
					) }
					description={ sprintf(
						/* translators: 1: Current GA4 property name. 2: Current GA4 property ID. 3: Newly linked GA4 property name. 4: Newly linked GA4 property ID. */
						__(
							'The Google Tag on your site is no longer associated with your current Google Analytics property "%1$s (%2$s)". It is now recording metrics to another Google Analytics property "%3$s (%4$s)". If you want to continue seeing Analytics data in the Site Kit dashboard, we suggest you update Site Kitâ€™s Google Analytics configuration to show data for the property used in your Google Tag.',
							'google-site-kit'
						),
						currentAnalyticsProperty.displayName,
						currentAnalyticsProperty._id,
						newAnalyticsProperty.displayName,
						newAnalyticsProperty._id
					) }
					ctaButton={ {
						label: __( 'Use new property', 'google-site-kit' ),
						onClick: updateToNewAnalyticsConfig,
						inProgress: isDoingSubmitChanges,
					} }
					dismissButton={ {
						label: __(
							'Keep existing property',
							'google-site-kit'
						),
						onClick: updateGoogleTagConfig,
					} }
					dismissOptions={ {
						expiresInSeconds: 1,
					} }
					dismissOnCTAClick
				/>
			</Notification>
		);
	}

	if ( newGoogleTagID ) {
		return (
			<Notification className="googlesitekit-publisher-win">
				<BannerNotification
					notificationID={ id }
					type={ TYPES.WARNING }
					title={ __(
						'Your Google tag configuration has changed',
						'google-site-kit'
					) }
					description={ sprintf(
						/* translators: 1: Currently set Google Tag ID. 2: Newly linked Google Tag ID. 3: Current GA4 property name. 4: Current GA4 property ID. */
						__(
							'The Google tag for your Google Analytics configuration has changed from %1$s to %2$s. To keep using your current Google Analytics property "%3$s (%4$s)", you need to configure Site Kit to place the new Google tag %2$s instead.',
							'google-site-kit'
						),
						currentAnalyticsConfig.googleTagID,
						newGoogleTagID,
						currentAnalyticsProperty.displayName,
						currentAnalyticsProperty._id
					) }
					ctaButton={ {
						label: sprintf(
							/* translators: %s: Newly linked Google Tag ID. */
							__( 'Update Google tag to %s', 'google-site-kit' ),
							newGoogleTagID
						),
						onClick: updateGoogleTagConfig,
					} }
					dismissOptions={ {
						expiresInSeconds: 1,
					} }
					dismissOnCTAClick
				/>
			</Notification>
		);
	}
}
