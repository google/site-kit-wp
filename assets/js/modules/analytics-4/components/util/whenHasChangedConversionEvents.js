/**
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';

/**
 * Higher-Order Component to render wrapped components when changed conversion events are detected.
 *
 * @since n.e.x.t
 *
 * @return {Function} Enhancing function.
 */
export default function whenHasChangedConversionEvents() {
	return ( WrappedComponent ) => {
		function WhenHasChangedConversionEvents( props ) {
			const hasChangedConversionEvents = useSelect( ( select ) => {
				const {
					hasNewConversionReportingEvents,
					hasLostConversionReportingEvents,
				} = select( MODULES_ANALYTICS_4 );

				return (
					hasNewConversionReportingEvents() ||
					hasLostConversionReportingEvents()
				);
			}, [] );

			if ( hasChangedConversionEvents ) {
				return <WrappedComponent { ...props } />;
			}

			const { WidgetNull = null } = props;
			return WidgetNull && <WidgetNull />;
		}

		return WhenHasChangedConversionEvents;
	};
}
