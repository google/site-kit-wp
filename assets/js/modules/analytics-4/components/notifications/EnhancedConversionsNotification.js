/**
 * Analytics EnhancedConversionsNotification component.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { createInterpolateElement } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { escapeURI } from '@/js/util/escape-uri';
import { CORE_SITE } from '@/js/googlesitekit/datastore/site/constants';
import { MODULES_ANALYTICS_4 } from '@/js/modules/analytics-4/datastore/constants';
import { TYPES } from '@/js/components/Notice/constants';
import LearnMoreLink from '@/js/googlesitekit/notifications/components/common/LearnMoreLink';
import NoticeNotification from '@/js/googlesitekit/notifications/components/layout/NoticeNotification';

export const ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS =
	'ecee-notification-analytics';

export default function EnhancedConversionsNotification( {
	id,
	Notification,
} ) {
	const documentationURL = useSelect( ( select ) =>
		select( CORE_SITE ).getDocumentationLinkURL(
			'enhanced-conversions-analytics'
		)
	);
	const ctaURL = useSelect( ( select ) => {
		const { getAccountID, getPropertyID, getServiceURL } =
			select( MODULES_ANALYTICS_4 );

		const accountID = getAccountID();
		const propertyID = getPropertyID();

		return getServiceURL( {
			path: escapeURI`/a${ accountID }p${ propertyID }/admin/datapolicies/datacollection`,
		} );
	} );

	return (
		<Notification>
			<NoticeNotification
				notificationID={ id }
				type={ TYPES.INFO_ALT }
				title={ __(
					'Boost your Analytics data with enhanced conversions',
					'google-site-kit'
				) }
				description={ createInterpolateElement(
					__(
						'Site Kit now supports enhanced conversions. This free feature helps you get a more complete and reliable count of your sales and leads from your website, even when people switch devices. To activate, turn on the setting for collecting user data in your Analytics account. <a />',
						'google-site-kit'
					),
					{
						a: (
							<LearnMoreLink
								id={ id }
								label={ __( 'Learn more', 'google-site-kit' ) }
								url={ documentationURL }
								hideExternalIndicator
							/>
						),
					}
				) }
				dismissButton={ {
					label: __( 'No thanks', 'google-site-kit' ),
				} }
				ctaButton={ {
					label: __( 'Go to Analytics', 'google-site-kit' ),
					href: ctaURL,
					dismissOnClick: true,
					external: true,
					hideExternalIndicator: true,
				} }
			/>
		</Notification>
	);
}

EnhancedConversionsNotification.propTypes = {
	id: PropTypes.string.isRequired,
	Notification: PropTypes.elementType.isRequired,
};
