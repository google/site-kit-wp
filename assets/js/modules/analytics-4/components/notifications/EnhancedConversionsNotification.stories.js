/**
 * EnhancedConversionsNotification component stories.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { withNotificationComponentProps } from '@/js/googlesitekit/notifications/util/component-props';
import EnhancedConversionsNotification, {
	ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS,
} from '@/js/modules/analytics-4/components/notifications/EnhancedConversionsNotification';

const NotificationWithComponentProps = withNotificationComponentProps(
	ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS
)( EnhancedConversionsNotification );

function Template() {
	return <NotificationWithComponentProps />;
}

export const Default = Template.bind( {} );
Default.storyName = 'Analytics';
Default.parameters = {
	features: [ 'gtagUserData' ],
};

export default {
	title: 'Components/Notifications/Subtle/EnhancedConversionsNotification',
	component: EnhancedConversionsNotification,
};
