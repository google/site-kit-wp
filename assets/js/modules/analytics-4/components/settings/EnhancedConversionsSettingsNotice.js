/**
 * Analytics EnhancedConversionsSettingsNotice component.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import {
	createInterpolateElement,
	useCallback,
	useEffect,
} from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useDispatch, useSelect } from 'googlesitekit-data';
import { useInView } from '@/js/hooks/useInView';
import useNotificationEvents from '@/js/googlesitekit/notifications/hooks/useNotificationEvents';
import { escapeURI } from '@/js/util/escape-uri';
import { CORE_SITE } from '@/js/googlesitekit/datastore/site/constants';
import { CORE_USER } from '@/js/googlesitekit/datastore/user/constants';
import { ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS } from '@/js/modules/analytics-4/components/notifications/EnhancedConversionsNotification';
import { MODULES_ANALYTICS_4 } from '@/js/modules/analytics-4/datastore/constants';
import LearnMoreLink from '@/js/googlesitekit/notifications/components/common/LearnMoreLink';
import Notice from '@/js/components/Notice';

export default function EnhancedConversionsSettingsNotice( {
	type = Notice.TYPES.INFO,
} ) {
	const id = ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS;

	const inView = useInView();
	const trackEvents = useNotificationEvents( id );

	const documentationURL = useSelect( ( select ) =>
		select( CORE_SITE ).getDocumentationLinkURL(
			'enhanced-conversions-analytics'
		)
	);
	const ctaURL = useSelect( ( select ) => {
		const { getAccountID, getPropertyID, getServiceURL } =
			select( MODULES_ANALYTICS_4 );

		const accountID = getAccountID();
		const propertyID = getPropertyID();

		return getServiceURL( {
			path: escapeURI`/a${ accountID }p${ propertyID }/admin/datapolicies/datacollection`,
		} );
	} );
	const isDismissed = useSelect( ( select ) =>
		select( CORE_USER ).isItemDismissed( id )
	);

	const { dismissItem } = useDispatch( CORE_USER );

	const handleCTAClick = useCallback( async () => {
		// Dismiss the notice when the CTA is clicked.
		await dismissItem( id );

		trackEvents.confirm();
	}, [ dismissItem, id, trackEvents ] );

	const handleDismiss = useCallback( async () => {
		await dismissItem( id );

		trackEvents.dismiss();
	}, [ dismissItem, id, trackEvents ] );

	// Track view event when notice comes into view.
	useEffect( () => {
		if ( inView ) {
			trackEvents.view();
		}
	}, [ inView, trackEvents ] );

	if ( isDismissed ) {
		return null;
	}

	return (
		<Notice
			type={ type }
			title={ __(
				'Boost your Analytics data with enhanced conversions',
				'google-site-kit'
			) }
			description={ createInterpolateElement(
				__(
					'Site Kit now supports enhanced conversions. This free feature helps you get a more complete and reliable count of your sales and leads from your website, even when people switch devices. To activate, turn on the setting for collecting user data in your Analytics account. <a />',
					'google-site-kit'
				),
				{
					a: (
						<LearnMoreLink
							id={ id }
							label={ __( 'Learn more', 'google-site-kit' ) }
							url={ documentationURL }
							hideExternalIndicator
						/>
					),
				}
			) }
			dismissButton={ {
				label: __( 'No thanks', 'google-site-kit' ),
				onClick: handleDismiss,
			} }
			ctaButton={ {
				label: __( 'Go to Analytics', 'google-site-kit' ),
				href: ctaURL,
				onClick: handleCTAClick,
			} }
		/>
	);
}

EnhancedConversionsSettingsNotice.propTypes = {
	type: PropTypes.oneOf( Object.values( Notice.TYPES ) ),
};
