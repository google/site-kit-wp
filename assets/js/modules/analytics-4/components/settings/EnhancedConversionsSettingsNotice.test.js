/**
 * EnhancedConversionsSettingsNotice component tests.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	fireEvent,
	render,
	waitFor,
} from '../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideSiteInfo,
	provideUserInfo,
} from '../../../../../../tests/js/utils';
import * as tracking from '@/js/util/tracking';
import { CORE_USER } from '@/js/googlesitekit/datastore/user/constants';
import { ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS } from '@/js/modules/analytics-4/components/notifications/EnhancedConversionsNotification';
import { MODULES_ANALYTICS_4 } from '@/js/modules/analytics-4/datastore/constants';
import { VIEW_CONTEXT_SETTINGS } from '@/js/googlesitekit/constants';
import EnhancedConversionsSettingsNotice from '@/js/modules/analytics-4/components/settings/EnhancedConversionsSettingsNotice';

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'EnhancedConversionsNotification', () => {
	let registry;

	const dismissItemEndpointRegExp = new RegExp(
		'^/google-site-kit/v1/core/user/data/dismiss-item'
	);

	beforeEach( () => {
		registry = createTestRegistry();

		provideSiteInfo( registry );
		provideUserInfo( registry );

		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '123456',
			propertyID: '654321',
		} );

		mockTrackEvent.mockClear();
	} );

	it( 'should render the notice', async () => {
		const { container, waitForRegistry } = render(
			<EnhancedConversionsSettingsNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should dismiss the notice when the dismiss button is clicked', async () => {
		fetchMock.postOnce( dismissItemEndpointRegExp, {
			body: [ ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS ],
		} );

		const { getByRole, waitForRegistry } = render(
			<EnhancedConversionsSettingsNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		const dismissButton = getByRole( 'button', { name: /no thanks/i } );

		fireEvent.click( dismissButton );

		await waitFor( () => {
			expect( fetchMock ).toHaveFetchedTimes( 1 );
			expect( fetchMock ).toHaveFetched( dismissItemEndpointRegExp );
			expect(
				registry
					.select( CORE_USER )
					.isItemDismissed(
						ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS
					)
			).toBe( true );
		} );
	} );

	it( 'should not render when the notification is dismissed', async () => {
		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedItems( [
				ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS,
			] );

		const { container, waitForRegistry } = render(
			<EnhancedConversionsSettingsNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toBeEmptyDOMElement();
	} );

	describe( 'event tracking', () => {
		const eventCategory = `${ VIEW_CONTEXT_SETTINGS }_${ ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS }`;

		it( 'should track an event when the notification is viewed', async () => {
			const { waitForRegistry } = render(
				<EnhancedConversionsSettingsNotice />,
				{
					registry,
					viewContext: VIEW_CONTEXT_SETTINGS,
				}
			);

			await waitForRegistry();

			expect( mockTrackEvent ).toHaveBeenCalledWith(
				eventCategory,
				'view_notification'
			);
		} );

		it( 'should track an event when the CTA button is clicked', async () => {
			fetchMock.postOnce( dismissItemEndpointRegExp, {
				body: [ ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS ],
			} );

			const { getByRole, waitForRegistry } = render(
				<EnhancedConversionsSettingsNotice />,
				{
					registry,
					viewContext: VIEW_CONTEXT_SETTINGS,
				}
			);

			await waitForRegistry();

			const ctaButton = getByRole( 'button', {
				name: /go to analytics/i,
			} );

			// Add click handler to prevent navigation.
			ctaButton.addEventListener( 'click', ( e ) => e.preventDefault() );

			fireEvent.click( ctaButton );

			await waitFor( () => {
				expect( mockTrackEvent ).toHaveBeenCalledWith(
					eventCategory,
					'confirm_notification'
				);
			} );
		} );

		it( 'should track an event when the dismiss button is clicked', async () => {
			fetchMock.postOnce( dismissItemEndpointRegExp, {
				body: [ ENHANCED_CONVERSIONS_NOTIFICATION_ANALYTICS ],
			} );

			const { getByRole, waitForRegistry } = render(
				<EnhancedConversionsSettingsNotice />,
				{
					registry,
					viewContext: VIEW_CONTEXT_SETTINGS,
				}
			);

			await waitForRegistry();

			const dismissButton = getByRole( 'button', {
				name: /no thanks/i,
			} );

			fireEvent.click( dismissButton );

			await waitFor( () => {
				expect( mockTrackEvent ).toHaveBeenCalledWith(
					eventCategory,
					'dismiss_notification'
				);
			} );
		} );

		it( 'should track an event when the learn more link is clicked', async () => {
			const { getByRole, waitForRegistry } = render(
				<EnhancedConversionsSettingsNotice />,
				{
					registry,
					viewContext: VIEW_CONTEXT_SETTINGS,
				}
			);

			await waitForRegistry();

			const learnMoreLink = getByRole( 'link', {
				name: /learn more/i,
			} );

			fireEvent.click( learnMoreLink );

			await waitFor( () => {
				expect( mockTrackEvent ).toHaveBeenCalledWith(
					eventCategory,
					'click_learn_more_link',
					undefined,
					undefined
				);
			} );
		} );
	} );
} );
