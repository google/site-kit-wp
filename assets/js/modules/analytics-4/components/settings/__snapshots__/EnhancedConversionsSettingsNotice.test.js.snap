// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics EnhancedConversionsNotification should render the notice 1`] = `
<div>
  <div
    class="googlesitekit-notice googlesitekit-notice--info"
  >
    <div
      class="googlesitekit-notice__icon"
    >
      <svg />
    </div>
    <div
      class="googlesitekit-notice__content"
    >
      <p
        class="googlesitekit-notice__title"
      >
        Boost your Analytics data with enhanced conversions
      </p>
      <p
        class="googlesitekit-notice__description"
      >
        Site Kit now supports enhanced conversions. This free feature helps you get a more complete and reliable count of your sales and leads from your website, even when people switch devices. To activate, turn on the setting for collecting user data in your Analytics account. 
        <a
          aria-label="Learn more (opens in a new tab)"
          class="googlesitekit-cta-link"
          href="https://sitekit.withgoogle.com/support/?doc=enhanced-conversions-analytics"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="googlesitekit-cta-link__contents"
          >
            Learn more
          </span>
        </a>
      </p>
    </div>
    <div
      class="googlesitekit-notice__action"
    >
      <button
        class="mdc-button mdc-button--tertiary"
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          No thanks
        </span>
      </button>
      <a
        class="mdc-button googlesitekit-notice__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
        href="https://accounts.google.com/accountchooser?continue=https%3A%2F%2Fanalytics.google.com%2Fanalytics%2Fweb%2F%23%2Fa123456p654321%2Fadmin%2Fdatapolicies%2Fdatacollection&Email=wapuu.wordpress%40gmail.com"
        role="button"
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          Go to Analytics
        </span>
      </a>
    </div>
  </div>
</div>
`;
